File: src/services/bom.service.ts
import { request } from './api';

// ── BOM CRUD ────────────────────────────────────────────────────────

export interface BomVO {
  id: number;
  bomNo: string;
  name: string;
  projectId: number;
  productCode?: string;
  productName?: string;
  version?: string;
  status: number;
  totalItems?: number;
  isLatest?: boolean;
}

export interface BomItemVO {
  id: number;
  bomId: number;
  parentId?: number | null;
  name: string;
  materialCode?: string;
  specification?: string;
  drawingNo?: string;
  material?: string;
  surfaceTreatment?: string;
  sourceType?: string;
  quantity: number;
  weight?: number;
  totalWeight?: number;
  unitOfMeasure?: string;
  isVirtual?: boolean;
  storageLocation?: string;
  isOptional?: boolean;
  remark?: string;
  level?: number;
  levelNo?: string;
  seqNo?: number;
  children?: BomItemVO[];
}

export async function createBom(data: {
  name: string;
  projectId: number;
  productCode: string;
  productName: string;
}): Promise<{ code: number; data: BomVO }> {
  return request.post('/boms', data);
}

export async function getBomById(id: number | string): Promise<{ code: number; data: BomVO }> {
  return request.get(`/boms/${id}`);
}

export async function getBomsByProject(projectId: number | string): Promise<{ code: number; data: BomVO[] }> {
  return request.get('/boms', { params: { projectId } });
}

// ── BOM Structure ───────────────────────────────────────────────────

/** GET /api/boms/{bomId}/structure — loads BOM item tree by BOM ID */
export async function getBomStructure(bomId: number | string): Promise<{ code: number; data: BomItemVO[] }> {
  return request.get(`/boms/${bomId}/structure`);
}

/** Add item to a specific BOM */
export async function createBomItem(bomId: number | string, data: Record<string, any>): Promise<{ code: number; data: BomItemVO }> {
  return request.post(`/boms/${bomId}/items`, data);
}

export async function updateBomItem(itemId: number, data: Record<string, any>): Promise<{ code: number; data: BomItemVO }> {
  return request.put(`/boms/items/${itemId}`, data);
}

export async function deleteBomItem(itemId: number): Promise<{ code: number; data: null }> {
  return request.delete(`/boms/items/${itemId}`);
}

// ── Approval ────────────────────────────────────────────────────────

/** Submit BOM for approval (BOM_APPROVAL workflow) */
export async function submitForApproval(bomId: number | string): Promise<{ code: number; data: null }> {
  return request.post(`/boms/${bomId}/submit-approval`);
}

/** Withdraw a pending BOM approval */
export async function withdrawBomApproval(bomId: number | string): Promise<{ code: number; data: null }> {
  return request.post(`/boms/${bomId}/withdraw-approval`);
}

// ── Versions ────────────────────────────────────────────────────────

export async function getBomVersions(id: number | string) {
  return request.get(`/boms/${id}/versions`);
}

/** Save current state as a new version. Backend: POST /api/boms/{id}/save-version?changeSummary=... */
export async function saveBomVersion(id: number | string, changeSummary?: string) {
  return request.post(`/boms/${id}/save-version`, null, { params: { changeSummary } });
}

/** Compare two BOM versions — returns added/removed/modified item lists */
export async function compareBomVersions(id: number | string, v1: number | string, v2: number | string) {
  return request.get(`/boms/${id}/compare`, { params: { v1, v2 } });
}

// ── Change Requests ─────────────────────────────────────────────────

export async function getChangeRequests(bomId: number | string) {
  return request.get(`/boms/${bomId}/change-requests`);
}

export async function createChangeRequest(bomId: number | string, data: {
  changeType: string;
  itemId?: number;
  name?: string;
  sourceType?: string;
  quantity?: number;
  specification?: string;
  materialCode?: string;
  unitOfMeasure?: string;
  description?: string;
}) {
  return request.post(`/boms/${bomId}/change-requests`, data);
}

// ── Legacy aliases (kept for backward compat) ───────────────────────

/** @deprecated Use getBomStructure(bomId) with BOM ID instead of project ID */
export const getBomTree = getBomStructure;

/** @deprecated Use getBomsByProject */
export const getBomItems = getBomsByProject;

/** @deprecated Use saveBomVersion */
export const createBomVersion = saveBomVersion;

/** Roll back a BOM to a previously saved version. Backend: POST /api/boms/{id}/rollback?targetVersion=... */
export async function rollbackBomVersion(id: number | string, targetVersion: string) {
  return request.post(`/boms/${id}/rollback`, null, { params: { targetVersion } });
}

File: syncflow-java/syncflow-bom/src/main/java/com/syncflow/bom/service/BomService.java
package com.syncflow.bom.service;

import com.syncflow.bom.dto.*;

import java.util.List;

/**
 * BOM management service interface.
 */
public interface BomService {

    /**
     * List all BOMs for the current tenant (paginated).
     */
    List<BomVO> listBoms(Long projectId);

    /**
     * Get BOM detail by id.
     */
    BomVO getBomDetail(Long id);

    /**
     * Create a new BOM.
     */
    BomVO createBom(CreateBomDTO dto);

    /**
     * Get the BOM item tree for a given BOM.
     */
    List<BomItemTreeVO> getBomStructure(Long bomId);

    /**
     * Add a BOM item to a BOM.
     */
    BomItemTreeVO addBomItem(Long bomId, CreateBomItemDTO dto);

    /**
     * Update a BOM item.
     */
    BomItemTreeVO updateBomItem(Long itemId, CreateBomItemDTO dto);

    /**
     * Delete a BOM item (and its children).
     */
    void deleteBomItem(Long itemId);

    /**
     * Submit a BOM for approval (status change + workflow trigger).
     */
    void submitForApproval(Long bomId);

    /**
     * Withdraw a pending BOM approval.
     */
    void withdrawApproval(Long bomId);

    /**
     * Save a new version snapshot of the BOM.
     */
    BomVO saveVersion(Long bomId, String changeSummary);

    /**
     * Get version history of a BOM.
     */
    List<BomVersionVO> getVersionHistory(Long bomId);

    /**
     * Compare two versions of a BOM and return item-level diffs.
     * <p>
     * {@code bomId} may be any BOM id in the version family (original or derived).
     * {@code v1} and {@code v2} are version strings such as "1.0" and "1.1".
     *
     * @throws com.syncflow.common.exception.BusinessException if the BOM is not found
     */
    BomVersionCompareVO compareVersions(Long bomId, String v1, String v2);

    /**
     * Roll back the BOM to a previously saved version.
     * <p>
     * Replaces the current items of {@code bomId} with items copied from the BOM
     * that owns the {@code targetVersion} record within the same version family.
     *
     * @throws com.syncflow.common.exception.BusinessException if the BOM or target version is not found
     */
    void rollbackVersion(Long bomId, String targetVersion);
}
